File: modules/quicksort/quicksort.go
// quicksort.go
//
// A Go port of the quicksort example for our DSA class.
// Adapted from the code provided with:
//   Java Foundations (2nd & 3rd ed) by Lewis, DePasquale, & Chase
//   Algorithms (4th ed) by Sedgewick & Wayne
//
// Run:
//   go run quicksort.go

package main

import (
	"fmt"
	"math/rand"
	"strings"
)

// ---- Helper operations -------------------------------------------------
// These operations occur multiple times in our sorting routines,
//   so we pull them out into small helpers.

func swap(data []int, ii, jj int) {
	data[ii], data[jj] = data[jj], data[ii]
}

func isSorted(data []int) bool {
	for ii := 1; ii < len(data); ii++ {
		if data[ii] < data[ii-1] {
			return false
		}
	}
	return true
}

// ---- Quicksort ---------------------------------------------------------

func partition(data []int, min, max int) int {
	middle := min + ((max - min) / 2)

	// Use the middle data value as the partition element,
	//   then move it out of the way (into the min slot) for now.
	partitionElement := data[middle]
	swap(data, middle, min)

	left := min
	right := max

	for left < right {
		// search for an element that is > the partition element
		for left < right && data[left] <= partitionElement {
			left++
		}

		// search for an element that is < the partition element
		for data[right] > partitionElement {
			right--
		}

		// swap the elements
		if left < right {
			swap(data, left, right)
		}
	}

	// move the partition element into place
	swap(data, min, right)

	return right
}

func quickSort(data []int, min, max int) {
	if min < max {
		// create partitions
		indexOfPartition := partition(data, min, max)

		// sort the left partition (lower values)
		quickSort(data, min, indexOfPartition-1)

		// sort the right partition (higher values)
		quickSort(data, indexOfPartition+1, max)
	}
}

// quickSortAll is the convenience entry point that sorts the whole slice.
func quickSortAll(data []int) {
	quickSort(data, 0, len(data)-1)
}

// ---- Test harness ------------------------------------------------------

func printArray(a []int) {
	parts := make([]string, len(a))
	for ii, value := range a {
		parts[ii] = fmt.Sprintf("%d", value)
	}
	fmt.Println(strings.Join(parts, " "))
}

func main() {
	failures := 0

	for kk := 0; kk < 5; kk++ {
		a := make([]int, 100)
		for ii := range a {
			a[ii] = rand.Intn(1000)
		}

		fmt.Print("\nUnsorted: ")
		printArray(a)

		quickSortAll(a)

		fmt.Print("  Sorted: ")
		printArray(a)

		if !isSorted(a) {
			fmt.Println("Fail!")
			failures++
		}
	}

	fmt.Println()
	if failures == 0 {
		fmt.Printf("Test successful! (%d failures)\n", failures)
	} else {
		fmt.Printf("Test unsuccessful! (%d failures)\n", failures)
	}
}

File: modules/quicksort/quick_sort.py
"""quick_sort.py

A Python port of the quicksort example for our DSA class.
Adapted from the code provided with:
  Java Foundations (2nd & 3rd ed) by Lewis, DePasquale, & Chase
  Algorithms (4th ed) by Sedgewick & Wayne

Run:
  python quick_sort.py
"""

import random


# ---- Helper operations -------------------------------------------------
# These operations occur multiple times in our sorting routines,
#   so we pull them out into small helpers.

def swap(data, ii, jj):
    data[ii], data[jj] = data[jj], data[ii]


def is_sorted(data, lo=0, hi=None):
    if hi is None:
        hi = len(data) - 1
    for ii in range(lo + 1, hi + 1):
        if data[ii] < data[ii - 1]:
            return False
    return True


# ---- Quicksort ---------------------------------------------------------

def partition(data, lo, hi):
    middle = lo + ((hi - lo) // 2)

    # Use the middle data value as the partition element,
    #   then move it out of the way (into the lo slot) for now.
    partition_element = data[middle]
    swap(data, middle, lo)

    left = lo
    right = hi

    while left < right:
        # search for an element that is > the partition element
        while left < right and data[left] <= partition_element:
            left += 1

        # search for an element that is < the partition element
        while data[right] > partition_element:
            right -= 1

        # swap the elements
        if left < right:
            swap(data, left, right)

    # move the partition element into place
    swap(data, lo, right)

    return right


def quick_sort(data, lo=0, hi=None):
    if hi is None:
        hi = len(data) - 1

    if lo < hi:
        # create partitions
        index_of_partition = partition(data, lo, hi)

        # sort the left partition (lower values)
        quick_sort(data, lo, index_of_partition - 1)

        # sort the right partition (higher values)
        quick_sort(data, index_of_partition + 1, hi)


# ---- Test harness ------------------------------------------------------

def print_array(a):
    print(" ".join(str(value) for value in a))


def main():
    failures = 0

    for _ in range(5):
        a = [random.randint(0, 999) for _ in range(100)]

        print("\nUnsorted: ", end="")
        print_array(a)

        quick_sort(a)

        print("  Sorted: ", end="")
        print_array(a)

        if not is_sorted(a):
            print("Fail!")
            failures += 1

    print()
    if failures == 0:
        print(f"Test successful! ({failures} failures)")
    else:
        print(f"Test unsuccessful! ({failures} failures)")


if __name__ == "__main__":
    main()
